### PR TITLE
substituteAll: disallow invalid arguments, fix plasma-workspace

### LINKED ISSUE #350037: plasma-workspace: substitution broken in plasma-ksplash-ready.service
### Describe the bug
`plasma-ksplash-ready.service: Unable to locate executable '-session': No such file or directory`

```
❯ systemctl --user cat plasma-ksplash-ready.service
# /run/current-system/sw/share/systemd/user/plasma-ksplash-ready.service
[Unit]
Description=KSplash "ready" Stage
Wants=plasma-core.target
After=plasma-core.target
PartOf=graphical-session.target

[Service]
Type=oneshot
ExecStart= --session --reply-timeout=1 --type=method_call --dest=org.kde.KSplash /KSplash org.kde.KSplash.setStage string:ready
Slice=session.slice
```

It causes KDE splash screen to take about 30 seconds until getting into the desktop.

### Steps To Reproduce
1. Install KDE plasma 6.2.1.1
2. Boot

### Expected behavior
The splash screen usually takes a few seconds to finish.

### Notify maintainers

@K900 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.11.4, NixOS, 24.11 (Vicuna), 24.11.20241020.eae12a7`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.9`
 - nixpkgs: `/nix/store/cg2ygk1w3mp1wmvi1p03d2lxyqnchs97-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by K900**
Fixes #350037, running a full eval to see if there's more.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/build-support/substitute/substitute-all.nix [hunk 1]
```diff
@@ -1,12 +1,20 @@
...
-  name = if args ? name then args.name else baseNameOf (toString args.src);
-  builder = ./substitute-all.sh;
-  inherit (args) src;
-  preferLocalBuild = true;
-  allowSubstitutes = false;
-} // args)
+args:
+let
+  # keep this in sync with substituteAll
+  isInvalidArgName = x: builtins.match "^[a-z][a-zA-Z0-9_]*$" x == null;
```
Comment (SuperSandro2000) [inline pkgs/build-support/substitute/substitute-all.nix:6]: At least in bash variable names can also start with underscores and can be one character long.

Comment (K900) [reply pkgs/build-support/substitute/substitute-all.nix:6]: That's the exact regex `substituteAll` uses.

Comment (SuperSandro2000) [reply pkgs/build-support/substitute/substitute-all.nix:6]: Then we shouldn't write below that it must be a valid shell variable name.

Comment (K900) [reply pkgs/build-support/substitute/substitute-all.nix:6]: Yeah, reworded the message to actually describe the requirements.

**Comment by spiage** [comment]
```
• Updated input 'nixpkgs':
    'github:nixos/nixpkgs/2a95839037749bf59b4457699190eeffa491b2fb?narHash=sha256-MKLMLahlSmN9uartRtQ2guCJiBYHTpWmZLgYBdbM5dg%3D' (2024-10-20)
  → 'github:nixos/nixpkgs/06224675bbdd09e378047904aa80462d637d5422?narHash=sha256-KmgnqlChyeF23RcYTKGx2h5pEPoZwnG1BT0qMl/DRcQ%3D' (2024-10-20)
$ cat /proc/sys/kernel/hostname
building the system configuration...
Building in flake mode.
$ nix --extra-experimental-features nix-command flakes build .#nixosConfigurations."a7".config.system.build.toplevel -v --keep-going --keep-failed --out-link /tmp/nixos-rebuild.3NHLQi/result
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:34:12:
           33|
           34|   strict = derivationStrict drvAttrs;
             |            ^
           35|

       … while evaluating derivation 'nixos-system-a7-24.11.20241020.0622467'
         whose name attribute is located at /nix/store/zl080npmvz2a1ng7c9mmkgx0qy3i02q7-source/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'buildCommand' of derivation 'nixos-system-a7-24.11.20241020.0622467'
         at /nix/store/zl080npmvz2a1ng7c9mmkgx0qy3i02q7-source/nixos/modules/system/activation/top-level.nix:59:5:
           58|     passAsFile = [ "extraDependencies" ];
           59|     buildCommand = systemBuilder;
             |     ^
           60|

       … while evaluating the option `system.systemBuilderCommands':

       … while evaluating definitions from `/nix/store/zl080npmvz2a1ng7c9mmkgx0qy3i02q7-source/nixos/modules/system/activation/activatable-system.nix':

       … while evaluating the option `system.activatableSystemBuilderCommands':

       … while evaluating definitions from `/nix/store/zl080npmvz2a1ng7c9mmkgx0qy3i02q7-source/nixos/modules/system/activation/switchable-system.nix':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: Argument names for `pkgs.substituteAll` must:
         - start with a lower case ASCII letter
         - only contain ASCII letters, digits and underscores
       Found invalid argument names: edk2-uefi-shell.
```

**Comment by spiage** [comment]
@k900 @SuperSandro2000

**Comment by K900** [comment]
https://github.com/NixOS/nixpkgs/pull/350081

**Pull request merged**